### ISSUE TITLE
How to implement car parking?

### ISSUE THREAD

**Issue by shenlonghu** [matsim-org/matsim-code-examples#226]
I am doing a project that 'how increase/decrease car-parking spaces in a campus will affect people's mode choice.' Can someone give me some advice about it?

**Comment by mrieser** [comment]
By default, MATSim assumes that agents using the car will find a parking space on their destination link. Thus, there is no default data structure for specifying car-parking in MATSim.

For car-parking spaces having an influence on the simulation, you would need to at least:
- keep track of used parking spaces
- have a custom scoring function that adds a penalty if there is no more parking space available.

More complex setups can be thought of, e.g. having agents drive further to look for an available parking space. But this would require a very good knowledge of MATSim to achieve, and might not replay give you better results than a simple "penalty" approach as above.

**Comment by shenlonghu** [comment]
Thanks for your reply! Much appreciate.

But i don't know how to keep track of used parking spaces. Could you please tell some details of this.

And i also read some literature which says to simulate car-parking, the original 'acticity-car-activity' became 'activity-walk-car interaction-car-parking-walk-activity'.

**Comment by kainagel** [comment]
There is a "parking" contrib, look at  the "parkingchoice" part of it.  E.g. see here:  https://github.com/matsim-org/matsim/blob/master/contribs/parking/src/main/java/org/matsim/contrib/parking/parkingchoice/package-info.java .  We have found that for our purposes (@ VSP) the assumptions were not the way we needed them.  In particular, not modelling the additional walk legs explicitly is problematic for us, and retrofitting it into the existing code seemed like a lot of work.  So we ended up rather programming it ourselves, in the way Marcel wrote. 

I am not sure if there is a maintainer of the parking contrib.
